### PR TITLE
P/brent/big endian bit mapping

### PR DESCRIPTION
**Pull request by brentpicasso**
So, We were doing can mapping for bit-mode big-endian incorrectly. 

Bits in a CAN message must be defined left to right, starting with the msb bit of byte 0. 

Arbitrary length values must be extracted from this 8 byte / 64 bit block before endian is converted. 

Below is the output of unit tests for a 9 bit extraction, to illustrate the desired effect. 
enable the #define CAN_MAPPING_TEST_DEBUG  in can_mapping_test.cpp to see the output. 

bitmode test: endian=1 / test_value=511 / offset=0 / length=9 / return = 511.000000
CAN data: ff 80 00 00 00 00 00 00 
11111111 10000000 00000000 00000000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=1 / length=9 / return = 511.000000
CAN data: 7f c0 00 00 00 00 00 00 
01111111 11000000 00000000 00000000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=2 / length=9 / return = 511.000000
CAN data: 3f e0 00 00 00 00 00 00 
00111111 11100000 00000000 00000000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=3 / length=9 / return = 511.000000
CAN data: 1f f0 00 00 00 00 00 00 
00011111 11110000 00000000 00000000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=4 / length=9 / return = 511.000000
CAN data: 0f f8 00 00 00 00 00 00 
00001111 11111000 00000000 00000000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=5 / length=9 / return = 511.000000
CAN data: 07 fc 00 00 00 00 00 00 
00000111 11111100 00000000 00000000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=6 / length=9 / return = 511.000000
CAN data: 03 fe 00 00 00 00 00 00 
00000011 11111110 00000000 00000000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=7 / length=9 / return = 511.000000
CAN data: 01 ff 00 00 00 00 00 00 
00000001 11111111 00000000 00000000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=8 / length=9 / return = 511.000000
CAN data: 00 ff 80 00 00 00 00 00 
00000000 11111111 10000000 00000000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=9 / length=9 / return = 511.000000
CAN data: 00 7f c0 00 00 00 00 00 
00000000 01111111 11000000 00000000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=10 / length=9 / return = 511.000000
CAN data: 00 3f e0 00 00 00 00 00 
00000000 00111111 11100000 00000000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=11 / length=9 / return = 511.000000
CAN data: 00 1f f0 00 00 00 00 00 
00000000 00011111 11110000 00000000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=12 / length=9 / return = 511.000000
CAN data: 00 0f f8 00 00 00 00 00 
00000000 00001111 11111000 00000000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=13 / length=9 / return = 511.000000
CAN data: 00 07 fc 00 00 00 00 00 
00000000 00000111 11111100 00000000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=14 / length=9 / return = 511.000000
CAN data: 00 03 fe 00 00 00 00 00 
00000000 00000011 11111110 00000000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=15 / length=9 / return = 511.000000
CAN data: 00 01 ff 00 00 00 00 00 
00000000 00000001 11111111 00000000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=16 / length=9 / return = 511.000000
CAN data: 00 00 ff 80 00 00 00 00 
00000000 00000000 11111111 10000000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=17 / length=9 / return = 511.000000
CAN data: 00 00 7f c0 00 00 00 00 
00000000 00000000 01111111 11000000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=18 / length=9 / return = 511.000000
CAN data: 00 00 3f e0 00 00 00 00 
00000000 00000000 00111111 11100000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=19 / length=9 / return = 511.000000
CAN data: 00 00 1f f0 00 00 00 00 
00000000 00000000 00011111 11110000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=20 / length=9 / return = 511.000000
CAN data: 00 00 0f f8 00 00 00 00 
00000000 00000000 00001111 11111000 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=21 / length=9 / return = 511.000000
CAN data: 00 00 07 fc 00 00 00 00 
00000000 00000000 00000111 11111100 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=22 / length=9 / return = 511.000000
CAN data: 00 00 03 fe 00 00 00 00 
00000000 00000000 00000011 11111110 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=23 / length=9 / return = 511.000000
CAN data: 00 00 01 ff 00 00 00 00 
00000000 00000000 00000001 11111111 00000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=24 / length=9 / return = 511.000000
CAN data: 00 00 00 ff 80 00 00 00 
00000000 00000000 00000000 11111111 10000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=25 / length=9 / return = 511.000000
CAN data: 00 00 00 7f c0 00 00 00 
00000000 00000000 00000000 01111111 11000000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=26 / length=9 / return = 511.000000
CAN data: 00 00 00 3f e0 00 00 00 
00000000 00000000 00000000 00111111 11100000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=27 / length=9 / return = 511.000000
CAN data: 00 00 00 1f f0 00 00 00 
00000000 00000000 00000000 00011111 11110000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=28 / length=9 / return = 511.000000
CAN data: 00 00 00 0f f8 00 00 00 
00000000 00000000 00000000 00001111 11111000 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=29 / length=9 / return = 511.000000
CAN data: 00 00 00 07 fc 00 00 00 
00000000 00000000 00000000 00000111 11111100 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=30 / length=9 / return = 511.000000
CAN data: 00 00 00 03 fe 00 00 00 
00000000 00000000 00000000 00000011 11111110 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=31 / length=9 / return = 511.000000
CAN data: 00 00 00 01 ff 00 00 00 
00000000 00000000 00000000 00000001 11111111 00000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=32 / length=9 / return = 511.000000
CAN data: 00 00 00 00 ff 80 00 00 
00000000 00000000 00000000 00000000 11111111 10000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=33 / length=9 / return = 511.000000
CAN data: 00 00 00 00 7f c0 00 00 
00000000 00000000 00000000 00000000 01111111 11000000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=34 / length=9 / return = 511.000000
CAN data: 00 00 00 00 3f e0 00 00 
00000000 00000000 00000000 00000000 00111111 11100000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=35 / length=9 / return = 511.000000
CAN data: 00 00 00 00 1f f0 00 00 
00000000 00000000 00000000 00000000 00011111 11110000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=36 / length=9 / return = 511.000000
CAN data: 00 00 00 00 0f f8 00 00 
00000000 00000000 00000000 00000000 00001111 11111000 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=37 / length=9 / return = 511.000000
CAN data: 00 00 00 00 07 fc 00 00 
00000000 00000000 00000000 00000000 00000111 11111100 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=38 / length=9 / return = 511.000000
CAN data: 00 00 00 00 03 fe 00 00 
00000000 00000000 00000000 00000000 00000011 11111110 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=39 / length=9 / return = 511.000000
CAN data: 00 00 00 00 01 ff 00 00 
00000000 00000000 00000000 00000000 00000001 11111111 00000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=40 / length=9 / return = 511.000000
CAN data: 00 00 00 00 00 ff 80 00 
00000000 00000000 00000000 00000000 00000000 11111111 10000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=41 / length=9 / return = 511.000000
CAN data: 00 00 00 00 00 7f c0 00 
00000000 00000000 00000000 00000000 00000000 01111111 11000000 00000000 
bitmode test: endian=1 / test_value=511 / offset=42 / length=9 / return = 511.000000
CAN data: 00 00 00 00 00 3f e0 00 
00000000 00000000 00000000 00000000 00000000 00111111 11100000 00000000 
bitmode test: endian=1 / test_value=511 / offset=43 / length=9 / return = 511.000000
CAN data: 00 00 00 00 00 1f f0 00 
00000000 00000000 00000000 00000000 00000000 00011111 11110000 00000000 
bitmode test: endian=1 / test_value=511 / offset=44 / length=9 / return = 511.000000
CAN data: 00 00 00 00 00 0f f8 00 
00000000 00000000 00000000 00000000 00000000 00001111 11111000 00000000 
bitmode test: endian=1 / test_value=511 / offset=45 / length=9 / return = 511.000000
CAN data: 00 00 00 00 00 07 fc 00 
00000000 00000000 00000000 00000000 00000000 00000111 11111100 00000000 
bitmode test: endian=1 / test_value=511 / offset=46 / length=9 / return = 511.000000
CAN data: 00 00 00 00 00 03 fe 00 
00000000 00000000 00000000 00000000 00000000 00000011 11111110 00000000 
bitmode test: endian=1 / test_value=511 / offset=47 / length=9 / return = 511.000000
CAN data: 00 00 00 00 00 01 ff 00 
00000000 00000000 00000000 00000000 00000000 00000001 11111111 00000000 
bitmode test: endian=1 / test_value=511 / offset=48 / length=9 / return = 511.000000
CAN data: 00 00 00 00 00 00 ff 80 
00000000 00000000 00000000 00000000 00000000 00000000 11111111 10000000 
bitmode test: endian=1 / test_value=511 / offset=49 / length=9 / return = 511.000000
CAN data: 00 00 00 00 00 00 7f c0 
00000000 00000000 00000000 00000000 00000000 00000000 01111111 11000000 
bitmode test: endian=1 / test_value=511 / offset=50 / length=9 / return = 511.000000
CAN data: 00 00 00 00 00 00 3f e0 
00000000 00000000 00000000 00000000 00000000 00000000 00111111 11100000 
bitmode test: endian=1 / test_value=511 / offset=51 / length=9 / return = 511.000000
CAN data: 00 00 00 00 00 00 1f f0 
00000000 00000000 00000000 00000000 00000000 00000000 00011111 11110000 
bitmode test: endian=1 / test_value=511 / offset=52 / length=9 / return = 511.000000
CAN data: 00 00 00 00 00 00 0f f8 
00000000 00000000 00000000 00000000 00000000 00000000 00001111 11111000 
bitmode test: endian=1 / test_value=511 / offset=53 / length=9 / return = 511.000000
CAN data: 00 00 00 00 00 00 07 fc 
00000000 00000000 00000000 00000000 00000000 00000000 00000111 11111100 
bitmode test: endian=1 / test_value=511 / offset=54 / length=9 / return = 511.000000
CAN data: 00 00 00 00 00 00 03 fe 
00000000 00000000 00000000 00000000 00000000 00000000 00000011 11111110 
bitmode test: endian=1 / test_value=511 / offset=55 / length=9 / return = 511.000000
CAN data: 00 00 00 00 00 00 01 ff 
00000000 00000000 00000000 00000000 00000000 00000000 00000001 11111111

### CODE REVIEW

File: src/util/byteswap.c [hunk 1]
```diff
@@ -49,3 +49,27 @@ int32_t swap_int32( int32_t val )
...
+{
+    return (val & 0x00000000000000FFUL) << 56 | (val & 0x000000000000FF00UL) << 40 |
+             (val & 0x0000000000FF0000UL) << 24 | (val & 0x00000000FF000000UL) << 8 |
+             (val & 0x000000FF00000000UL) >> 8 | (val & 0x0000FF0000000000UL) >> 24 |
+             (val & 0x00FF000000000000UL) >> 40 | (val & 0xFF00000000000000UL) >> 56;
+}
+
+uint32_t swap_uint_length(uint32_t val, size_t bit_length)
+{
+    if (bit_length > 8) {
```
Comment (stieg) [inline src/util/byteswap.c]: Also why no support for uint64 here?

File: test/can_obd2/can_mapping_test.cpp [hunk 2]
```diff
@@ -72,85 +76,128 @@ void CANMappingTest::formula_test(void)
...
-                    mapping.offset = offset;
-                    mapping.length = length;
-                    mapping.type = CANMappingType_unsigned;
-                    mapping.bit_mode = true;
-                    msg.data64 = test_value << offset;
-
-                    float value = canmapping_extract_value(msg.data64, &mapping);
+        srand(time(NULL));
+
+        for (size_t random = 0; random <= 1; random++){
```
Comment (stieg) [inline test/can_obd2/can_mapping_test.cpp]: So randomness in unit-testing is traditionally not so good usually because of the non-deterministic nature.  Are there some edge cases you can think of to test instead (like all 1's or all 0's).  I would focus on those more since it will be very difficult to ever debug a random failure.  But I like where your head is at here :1st_place_medal:

File: test/can_obd2/can_mapping_test.cpp [hunk 3]
```diff
@@ -72,85 +76,128 @@ void CANMappingTest::formula_test(void)
...
-                    printf("\r\n");
+                                        float value = canmapping_extract_value(msg.data64, &mapping);
 
+                                        /* prepare the comparison value */
+                                        uint64_t compare_value = test_value;
+                                        if (!mapping.big_endian) {
+                                                compare_value = swap_uint_length(test_value, length);
+                                        }
+#ifdef CAN_MAPPING_TEST_DEBUG
+                                        printf("bitmode test: endian=%u / test_value=%lu / offset=%d / length=%d / return = %f\r\n" ,
```
Comment (stieg) [inline test/can_obd2/can_mapping_test.cpp]: Perhaps get rid of this code?

File: src/util/byteswap.c (hunk 1 above)
Comment (brentpicasso) [inline src/util/byteswap.c]: good point, will add 64

I thought all one-line if statements did not comply to the style guide, but happy to make that change. 

-Brent

File: test/can_obd2/can_mapping_test.cpp (hunk 3 above)
Comment (brentpicasso) [inline test/can_obd2/can_mapping_test.cpp]: I've wanted to keep that for development purposes if/when it needs to be debug'd. So, I prefer keeping it.

File: test/can_obd2/can_mapping_test.cpp (hunk 2 above)
Comment (brentpicasso) [inline test/can_obd2/can_mapping_test.cpp]: agreed. I was having those feelings of uncertainty about the randomness. Could be good for a fuzzing test, but out of scope here.  I changed it to a bitpattern of 0x55 as an alternative to the all 1's.

File: test/can_obd2/can_mapping_test.cpp (hunk 3 above)
Comment (stieg) [inline test/can_obd2/can_mapping_test.cpp]: I get it.  Suggest using the macros we have for debug level printing.  I forget what they are off-hand but they exist and are in use.

File: src/util/byteswap.c [hunk 4]
```diff
@@ -58,18 +58,11 @@ uint64_t swap_uint64(uint64_t val)
...
-            else if (bit_length <= 24) {
-                    return swap_uint24(val);
-            }
-            else {
-                    return swap_uint32(val);
-            }
-    }
-    return val;
+    if (bit_length <= 8) return val;
+    if (bit_length <= 16) return swap_uint16(val);
```
Comment (stieg) [inline src/util/byteswap.c]: If these 1 line if statements don't comply with style guide then make them 2 line.  The logic here is what is most important and easier to understand; style less so.

File: src/util/byteswap.c [hunk 5]
```diff
@@ -58,18 +58,11 @@ uint64_t swap_uint64(uint64_t val)
...
+    if (bit_length <= 24) return swap_uint24(val);
+    if (bit_length <= 32) return swap_uint32(val);
+    return swap_uint64(val);
```
Comment (stieg) [inline src/util/byteswap.c]: You did not include the `panic`.  Should be this:
```c
if (bit_length <= 64)
   return swap_uint64(val);
panic(UNREACHABLE);
```

Comment (brentpicasso) [inline src/util/byteswap.c]: :+1:  fixed in https://github.com/autosportlabs/RaceCapture-Pro_firmware/commit/53999e73ad3ce037f9030edc0226048eb4381454

Comment (brentpicasso) [inline src/util/byteswap.c]: ...and https://github.com/autosportlabs/RaceCapture-Pro_firmware/commit/90ee1f239ab122eecb85659d794be548c1bc6767
